calico/openstack/t_etcd.py

```diff
@@ -67,19 +69,25 @@ def initialize(self):
         self.client = etcd.Client(host=cfg.CONF.calico.etcd_host,
                                   port=cfg.CONF.calico.etcd_port)
 
-        # Spawn a green thread for periodically resynchronizing etcd against
-        # the OpenStack database.
-        eventlet.spawn(self.periodic_resync_thread)
-
         # We will remember the OpenStack security group data, between resyncs,
         # so that we can generate profiles when needed for new or updated
         # endpoints.  Start off with an empty set.
         self.sgs = {}
+        # Semaphore to protect all access to the sgs dict.
+        self._sgs_semaphore = Semaphore(1)
```

Comment:
This semaphore is unnecessary.

eventlet is like gevent, it's a co-operative concurrency framework with implicit yielding. This means that a semaphore is only useful if it is unacceptable to see partial state of an object. That does not apply in this case.

There are only three accesses of the `sgs` dict in this code. One is a write into the dict in `security_group_updated`: this is a single write, so the flow of control will not be yielded and therefore it cannot conflict with anything.

Another is a read from the dict in `profile_rules`. This is protected by a retry, so it tolerates a partially complete dictionary, and therefore does not require the semaphore (it will yield if the dictionary element it's looking for is not present).

The last is the delete-and-write in `resync_security_groups`. The only reason to have a semaphore cover this operation is to avoid anyone seeing the partially populated (or empty) dictionary while we make DB queries, but given that neither of the other entry points is going to be perturbed by this it's hard to see why the semaphore is necessary here, also.

The only reason I can see for the semaphore is if we believe that a security group resynchronization is going to take more than 20 * 200ms (4 seconds). Do we believe that's a likely scenario?